cluster/store.go

```diff
@@ -641,3 +647,82 @@ func lastSnapshotIndex(ss *raft.FileSnapshotStore) uint64 {
 	}
 	return ls[0].Index
 }
+
+// recoverSingleNode is used to manually force a new configuration in order to
+// recover from a loss of quorum where the current configuration cannot be
+// WARNING! This operation implicitly commits all entries in the Raft log, so
+// in general this is an extremely unsafe operation and that's why it's made to be
+// used in a single cluster node.
+// for more details see : https://github.com/hashicorp/raft/blob/main/api.go#L279
+func (st *Store) recoverSingleNode() error {
+	if st.cfg.BootstrapExpect > 1 || len(st.candidates) > 1 {
+		return fmt.Errorf("bootstrap expect  %v, candidates %v,"+
+			"can't perform auto recovery in multi node cluster", st.cfg.BootstrapExpect, st.candidates)
+	}
+	servers := st.raft.GetConfiguration().Configuration().Servers
+	// nothing to do here, wasn't a single node
+	if len(servers) != 1 {
+		return nil
+	}
+
+	exNode := servers[0]
+	newNode := raft.Server{
+		ID:       raft.ServerID(st.cfg.NodeID),
+		Address:  raft.ServerAddress(fmt.Sprintf("%s:%d", st.cfg.Host, st.cfg.RPCPort)),
+		Suffrage: raft.Voter,
+	}
+
+	// same node nothing to do here
+	if exNode.ID == newNode.ID && exNode.Address == newNode.Address {
+		return nil
+	}
+
+	st.log.WithFields(logrus.Fields{
+		"action":                      "raft_cluster_recovery",
+		"existed_single_cluster_node": exNode,
+		"new_single_cluster_node":     newNode,
+	}).Info("perform cluster recovery")
+
+	fut := st.raft.Shutdown()
+	if err := fut.Error(); err != nil {
+		return err
+	}
+
+	if err := raft.RecoverCluster(st.raftConfig(), &Store{
+		cfg:           st.cfg,
+		log:           st.log,
+		raftResolver:  st.raftResolver,
+		raftTransport: st.raftTransport,
+		applyTimeout:  st.applyTimeout,
+		snapshotStore: st.snapshotStore,
+		schemaManager: st.schemaManager,
+		logStore:      st.logStore,
+		logCache:      st.logCache,
+	}, st.logCache,
+		st.logStore,
+		st.snapshotStore,
+		st.raftTransport,
+		raft.Configuration{Servers: []raft.Server{newNode}}); err != nil {
+		return err
+	}
+
+	var err error
+	st.raft, err = raft.NewRaft(st.raftConfig(), st, st.logCache, st.logStore, st.snapshotStore, st.raftTransport)
```

Comment:
Question: `RecoverCluster` doc advises to provide a "fresh" FSM after a call to `RecoverCluster`. I see that we're instantiating a `Store` just for the call so we avoid any "broken" state from that. However are we OK keeping the original `st` we tried to start already or do we want to clear `st` first to be safe ?

Reply:
having newRaft called with `st` will be enough w.r.t https://github.com/hashicorp/raft/blob/a5db400428964844d4a148da83d17a192ced345f/api.go#L577 . i think it shouldn't be concern given that we use that for single node cluster only